kong/runloop/plugins_iterator.lua

```diff
@@ -169,174 +147,105 @@ local function get_plugin_config(plugin, name, ws_id)
-
-  local route    = ctx.route
-  local service  = ctx.service
-  local consumer = ctx.authenticated_consumer
-
-  if route and plugin.handler.no_route then
-    route = nil
-  end
-  if service and plugin.handler.no_service then
-    service = nil
-  end
-  if consumer and plugin.handler.no_consumer then
-    consumer = nil
-  end
-
-  local    route_id = route    and    route.id or nil
-  local  service_id = service  and  service.id or nil
-  local consumer_id = consumer and consumer.id or nil
-
-  if kong.db.strategy == "off" then
-    if route_id and service_id and consumer_id and combos[COMBO_RSC]
-      and combos.rsc[route_id] and combos.rsc[route_id][service_id]
-      and combos.rsc[route_id][service_id][consumer_id]
-    then
-      return combos.rsc[route_id][service_id][consumer_id]
-    end
-
-    if route_id and consumer_id and combos[COMBO_RC]
-      and combos.rc[route_id] and combos.rc[route_id][consumer_id]
-    then
-      return combos.rc[route_id][consumer_id]
-    end
-
-    if service_id and consumer_id and combos[COMBO_SC]
-      and combos.sc[service_id] and combos.sc[service_id][consumer_id]
-    then
-      return combos.sc[service_id][consumer_id]
-    end
-
-    if route_id and service_id and combos[COMBO_RS]
-      and combos.rs[route_id] and combos.rs[route_id][service_id]
-    then
-      return combos.rs[route_id][service_id]
-    end
-
-    if consumer_id and combos[COMBO_C] and combos.c[consumer_id] then
-      return combos.c[consumer_id]
+---
+-- Lookup a configuration for a given combination of route_id, service_id, consumer_id
+--
+-- The function checks various combinations of route_id, service_id and consumer_id to find
+-- the best matching configuration in the given 'combos' table. The priority order is as follows:
+--
+-- Route, Service, Consumer
+-- Route, Consumer
+-- Service, Consumer
+-- Route, Service
+-- Consumer
+-- Route
+-- Service
+-- Global
+--
+-- @function lookup_cfg
+-- @tparam table combos A table containing configuration data indexed by compound keys.
+-- @tparam string|nil route_id The route identifier.
+-- @tparam string|nil service_id The service identifier.
+-- @tparam string|nil consumer_id The consumer identifier.
+-- @return any|nil The configuration corresponding to the best matching combination, or 'nil' if no configuration is found.
+---
+function PluginsIterator.lookup_cfg(combos, route_id, service_id, consumer_id)
+  -- Use the build_compound_key function to create an index for the 'combos' table
+  local build_compound_key = PluginsIterator.build_compound_key
+
+    local key
+    if route_id and service_id and consumer_id then
+        key = build_compound_key(route_id, service_id, consumer_id)
+        if combos[key] then
+            return combos[key]
+        end
     end
-
-    if route_id and combos[COMBO_R] and combos.r[route_id] then
-      return combos.r[route_id]
+    if route_id and consumer_id then
+        key = build_compound_key(route_id, nil, consumer_id)
```

Comment:
If service_id is `nil`, should it equal to `build_compound_key(route_id, service_id, consumer_id)`?

Why do we use such `if` condition? (sorry, I have not too many backgroud.)

Reply:
I got the idea.